Rectangle fast path tweaks for softgpu

unknownbrackets opened the pull request:
A few rectangle fast path related changes:
 * Fixes a bug where some verts would incorrectly be culled after a rectangle fast path, affecting FF4 menus.
 * Uses threads for fast path rectangles, which helped a good bit in some areas.
 * Detects triangle fans used to draw a trivial rectangle and sends to fast path, which helped Legend of Heroes III noticeably.
 * Allows fast path for non-textured rectangles (sometimes used for fills with clear off.)
 * Requires flat Z for fast path rectangles.
 * Corrects some range cull handling with fast paths.
 * Combines sliced rectangles (mainly clears) into one rectangle.

Personally, I think threading rectangles is a much smarter case than triangles, because they've got even width in our fast path cases.  Seeing 90 FPS -> 120 FPS in some areas of a few games.

-[Unknown]

unknownbrackets committed:
softgpu: Fix leaked range flag on cull.

Fixes some backgrounds in Final Fantasy 4, probably others.

unknownbrackets committed:
softgpu: Fast path triangles without textures.

The fast path may still be useful in this case.

unknownbrackets committed:
softgpu: Check flat z in fast path.

unknownbrackets committed:
softgpu: Use range loops for sprite fast path.

unknownbrackets committed:
softgpu: Fast path rectangles as fans.

Some games, such as Legend of Heroes III, use fans instead of strips.

unknownbrackets committed:
softgpu: Fixup range cull for fans and fast path.

unknownbrackets committed:
softgpu: Split out rectangle path for combining.

unknownbrackets committed:
softgpu: Combine sliced rectangles.

This mostly affects clears, and reduces overhead.  Only about 2%
improvement, but it's a small change.

unknownbrackets committed:
sfotgpu: Use threads to apply clears.

LunaMoo commented:
This glitches that one game where software renderer is forced(tried with and without the compat hack), here's a dump to reproduce easily.
[ULUS10005_0001.zip](https://github.com/hrydgard/ppsspp/files/7535444/ULUS10005_0001.zip)

unknownbrackets commented:
Oops, silly mistake, missed y1/y2 in one of the threading loops.  Thanks for the frame dump.

-[Unknown]

unknownbrackets committed:
softgpu: Tune thread minimums better.

Darkstalkers seems more sensitive to these than many other games, this
improves performance more.

unknownbrackets commented:
Also with that dump, I compared thread minimums a bit and now Darkstalkers seems to be faster as well (it got a bit slower before the last commit.)

-[Unknown]

Gamemulatorer commented:
This can help my issue https://github.com/hrydgard/ppsspp/issues/15062?

LunaMoo commented:
@unknownbrackets in-game during fight it's roughtly the same, but on some screens the difference got pretty huge like from 500 to 800 fps. :o Nice.

@Gamemulatorer doubt this affects your issue or even change the performance of that game much, but why do you even use software renderer to play Hakuto? Wasn't it fixed? If it still doesn't work with hardware renderer in master or is limited to backend you can't use, you can try my old cheat ~ https://github.com/LunaMoo/PPSSPP_workarounds

hrydgard commented on GPU/Software/Rasterizer.cpp:
```diff
@@ -1451,26 +1451,35 @@ void ClearRectangle(const VertexData &v0, const VertexData &v1)
-				for (int x = 0; x < w; ++x) {
-					SetPixelDepth(p.x + x, p.y, z);
+		if ((z & 0xFF) == (z >> 8)) {
+			ParallelRangeLoop(&g_threadManager, [=](int y1, int y2) {
```
I find it a little risky to use coordinates multiplied by 16 as range inputs to ParallelRangeloop, it might get the idea to divide by some odd number of CPUs and you might end up some kind of fractional lines here? Not sure if an actual issue though..

unknownbrackets replied:
Since the min thread size is a multiple of 16, this can't happen.

-[Unknown]

hrydgard replied:
I don't see how it can't happen. Let's say it's a 48 pixel high rectangle and it gets split into two tasks, with 24 entries each? At least we end up relying on details of the splitting algorithm..

unknownbrackets replied:
Okay, I divided it out so there's no concern of details.

-[Unknown]

hrydgard replied:
thanks! I think the code is slightly clearer now too.

hrydgard commented on GPU/Software/Rasterizer.cpp:
```diff
@@ -1451,26 +1451,35 @@ void ClearRectangle(const VertexData &v0, const VertexData &v1)
 	minY = std::max(minY, (int)TransformUnit::DrawingToScreen(scissorTL).y);
 	maxY = std::max(0, std::min(maxY, (int)TransformUnit::DrawingToScreen(scissorBR).y + 16));
 
+	constexpr int MIN_LINES_PER_THREAD = 32 * 16;
```
Somewhere around here, we should add a very brief comment explaining the multiplication by 16 to get subpixel precision.

unknownbrackets replied:
Okay, added a comment.

-[Unknown]

unknownbrackets committed:
softgpu: Add some comments and cleanup.

unknownbrackets committed:
softgpu: Clear by dividing out subpixel first.
